documentation/writing_tests.md

```diff
@@ -0,0 +1,1534 @@
+
+```ts
+  // `isRecordMode` and `isPlaybackMode` are methods exported from the Recorder.
+  if (!isRecordMode() && !isPlaybackMode()) {
+    return this.skip();
+  }
+```
+
+For example, in `@azure/keyvault-keys` we provide a method to purge keys that has not been moved to use `core-lro`. To check that it has finished, we do a while loop where we try to make the next operation until it passes:
+
+```ts
+import { delay, isRecordMode, isPlaybackMode } from "@azure/test-utils-recorder";
+// ...
+
+describe("Keys client - restore keys and recover backups", () => {
+  // ...
+  it("can restore a key with a given backup", async function() {
+
+    // This test can't be expected to finish in any reasonable time.
+    if (!isRecordMode() && !isPlaybackMode()) {
+      return this.skip();
+    }
+
+    // ...
+    await client.createKey(keyName, "RSA");
+    const backup = await client.backupKey(keyName);
+    await client.purgeDeletedKey(keyName);
+    while (true) {
+      try {
+        await client.restoreKeyBackup(backup as Uint8Array);
+        break;
+      } catch (e) {
+        console.log("Can't restore the key since it's not fully deleted:", e.message);
+        console.log("Retrying in one second...");
+        // This delay method comes from the recorder
+        await delay(1000);
+      }
+    }
+    const getResult = await client.getKey(keyName);
+    assert.equal(getResult.name, keyName, "Unexpected key name in result from getKey().");
+    // ...
+  });
+  // ...
+});
+```
+
+Keep in mind that Mocha will have a timeout configuration that will prevent this to run up forever.
+
+The specific `delay` method used in the code above comes from [the recorder](#the-recorder), so that in playback, there will be no delay at all, and tests will pass as soon as possible.
+
+### Exceptions and edge cases
+
+While testing the Azure SDK clients for JavaScript and TypeScript, we should document all known exceptions through the use of `@throws` in the TypeDoc documentation. Tests for exceptions should only exist if these exceptions have been implemented within the code of the SDK clients, and not if they happen on the service side.
+
+Similarly, since the public API surface of our clients contain a large set of properties resulting from any of the methods that our clients implement, our tests should not focus on verifying that all of the client properties exist or are expectedly received. Tests should include only as many properties as it can be relevant for the use case that each test case is representing. For this purpose we should also take advantage of strict types. If our types can be descriptive and thorough (, and our internal code is not skipping any types by using `any`), we will be able to trust that our API is behaving reasonably well.
```

Comment:
```suggestion
Similarly, since the public API surface of our clients contain a large set of properties resulting from any of the methods that our clients implement, our tests should not focus on verifying that all of the client properties exist or are expectedly received. Tests should include only as many properties as it can be relevant for the use case that each test case is representing. For this purpose we should also take advantage of strict types. If our types can be descriptive and thorough (and our internal code is not skipping any types by using `any`) we will be able to trust that our API is behaving reasonably well.
```